### DISCUSSION TITLE
Testing bloat from tower animations

### DISCUSSION THREAD

**Discussion by Praytic** [Praytic/youtd2#349] [General]
# Intro

Before creating costly animations for towers, we need to conduct profiling based on the estimations. From the results, we will then decide how to proceed. The initial concern is that animations for the large number of towers might lead to "bloat" in the game. This bloat could potentially affect several areas:

* Binary size
* Load time
* VRAM (Video RAM)
* FPT (Frame Processing Time)

# Estimation

## Sprites Count

Tower sprite sheets reuse visuals per 2 tiers (second visual is a scaled version of the first visual). That's why we have less **unique sprites** per tower family.

| Rarity    | Total Towers | Families | Unique Sprites |
|----------------|--------------|----------|----------------|
| **All**      | 666 towers   | 227      | 420            |
| Common  | 215 towers   | 40       | 120            |
| Uncommon | 228 towers   | 54       | 122            |
| Rare    | 140 towers   | 53       | 98             |
| Unique  | 83 towers    | 80       | 80             |

### Inanimate and Animate Models

We divide tower models into 2 groups: the first group consists of inanimate models, and the second group consists of animate models. For inanimate models, most animations can be done using particles, whereas for animate models, most animations require frame-by-frame animation of the sprite. Below is a table showing number of inanimate and animate unique sprites per tower rarity.

| Rarity    | Inanimate    |  Animate  | 
|----------------|--------------|--------------|
| **All**      | 242      | 178            |
| Common  |  117       | 3            |
| Uncommon | 67       |  53            |
| Rare    | 46       | 42             |
| Unique  | 0       | 80             |

The **animate** column will be used in optimistic estimation, because on the current stage we can't tell for sure which models will require frame-by-frame animation.

### Animation Variants

Towers should have 2 animation variants: **idle** and **attack**. Idle animation for animate models should be played with some periodicity to simulate breathing, inanimate idle animations also can be played with some interval to simulate 3D form of the tower (e.g. shimmering particles behind and in front of the tower sprite). Some inanimate towers require constant animation, like fire burning or flying orb around the tower.

Let's consider that inanimate models don't require frame-by-frame animation and it's enough to add particles with procedural animations for them. For animate models, idle animation duration can play up to 2 seconds with periodicity of 2 seconds (supposedly it can also be done with procedural animation). Attack animation for inanimate models should not require frame-by-frame animation in most cases, while for animate models it's required. Attack animation should be up to 2 seconds long.

All animations will be played at max 30 frames per second. We don't want flickering animations at the start of the game due to their low attack speed. This leaves us with 60 frames per animation variant. 

# Profiling

performed by @Kvel2D

I created 500 duplicate spritesheets based on existing "boss stun" animation. Each spritesheet has 56 frames which is close enough to 60 frames. 
Note that each frame is 140x140 and currently some tower sprites are 300x300, so this test is a *lower bound*.

![2024-01-12 07_39_46-Window](https://github.com/Praytic/youtd2/assets/13269470/42d7fd43-56f4-4f41-b66b-a27cc1bab3e6)

Effect on load time = no change

Effect on binary size = +500MB

![2024-01-12 08_14_50-Window](https://github.com/Praytic/youtd2/assets/13269470/76f8c115-a462-4bc6-9f78-754f69c5af73)

## VRAM usage

VRAM usage is harder to estimate because the game only loads tower scenes which exist in the game.

So all 500 tower animations will never be loaded in VRAM.

Let's say player builds about 100 towers over the course of the game, including upgrades.

VRAM consumption per tower is around 5MB:

![2024-01-12 07_53_46-Window](https://github.com/Praytic/youtd2/assets/13269470/bd096a52-c2c1-46fd-a123-66fc5f74df44)
![2024-01-12 08_02_25-Window](https://github.com/Praytic/youtd2/assets/13269470/336b0e40-8999-42c7-9b1f-7115ec66a398)

Did a rough test where I added an animation to a tower with 54 frames.

So total VRAM usage from tower animations = 100 * 5MB = 500MB

Currently, the game uses 500MB VRAM usage so total VRAM usage with tower animations would be 1GB.

# Summary

Note that these results assume that tower animations have been properly packed. You can expect unpacked spritesheets to worsen the bloat by 4-5 times.

* Binary size = +0.5GB
* Load time = no effect
* VRAM = +0.5GB

**Comment by sipabam** [comment]
can you help me with that!!
python how to make animation?

**Reply by Kvel2D** [reply]
Hello, you can find instructions here: https://github.com/Praytic/youtd2/wiki/Turning-3D-models-into-2D-sprites
